lib/models/bucket.dart

```diff
@@ -7,55 +6,58 @@ import 'package:vikunja_app/models/user.dart';
 @JsonSerializable()
 class Bucket {
   int id, listId, limit;
-  String? title;
-  double position;
-  DateTime? created, updated;
-  User? createdBy;
+  String title;
+  double? position;
+  late final DateTime created, updated;
+  User createdBy;
   bool isDoneBucket;
+  late final List<Task> tasks;
+
   Bucket({
-    required this.id,
+    this.id = -1,
     required this.listId,
-    this.title,
-    this.position = 0,
+    required this.title,
+    this.position,
     required this.limit,
     this.isDoneBucket = false,
-    this.created,
-    this.updated,
-    this.createdBy,
-    this.tasks = const <Task>[],
-  });
-
-  List<Task> tasks = [];
+    DateTime? created,
+    DateTime? updated,
+    required this.createdBy,
+    List<Task>? tasks,
+  }) {
+    this.created = created ?? DateTime.now();
+    this.updated = created ?? DateTime.now();
+    this.tasks = tasks ?? [];
+  }
 
   Bucket.fromJSON(Map<String, dynamic> json)
-    : id = json['id'],
-      listId = json['list_id'],
-      title = json['title'],
-      position = json['position'] is int
-          ? json['position'].toDouble()
-          : json['position'],
-      limit = json['limit'],
-      isDoneBucket = json['is_done_bucket'],
-      created = DateTime.parse(json['created']),
-      updated = DateTime.parse(json['updated']),
-      createdBy = json['created_by'] == null
-        ? null
-        : User.fromJson(json['created_by']),
-      tasks = (((json['tasks']  == null) ? [] : json['tasks']) as List<dynamic>)
-        .map((task) => Task.fromJson(task))
-        .cast<Task>()
-        .toList();
+      : id = json['id'],
+        listId = json['list_id'],
+        title = json['title'],
+        position = json['position'] is int
+            ? json['position'].toDouble()
+            : json['position'],
+        limit = json['limit'],
+        isDoneBucket = json['is_done_bucket'],
+        created = DateTime.parse(json['created']),
+        updated = DateTime.parse(json['updated']),
+        createdBy = User.fromJson(json['created_by']),
+        tasks = json['tasks'] == null
+            ? []
+            : (json['tasks'] as List<dynamic>)
+                .map((task) => Task.fromJson(task))
+                .toList();
 
   toJSON() => {
-    'id': id,
+    'id': id != -1 ? id : null,
```

Comment:
Did this this way, here and in other models, in order to make `id` non-nullable while also allowing new items to have a null id when the request is made to the server to create it. This allows the server to assign a new id.

Reply:
So why not make the id nullable in the first place?

Reply:
So why not make the id nullable in the first place?

Reply:
Because then we have to do null checks everywhere we want to use id, which is a lot of places. Also, you had it has non-nullable, but required and setting it to 0 when creating a new item. This just formalizes that with a default value instead. However, it does make sense semantically for id to be nullable to represent an item that hasn't had an id assigned to it yet. So, I'm not sure what we should do. I'm not sure what happens when give the Vikunja API a id <= 0, but I sort-of assumed that it would return an error.

Eventually, when we get around to adding offline capability, we will have to look at how the API calculates the id and do that too, so that items can be added to newly created items while offline and then do some sort of migration when coming back online. That's going to be really complicated.

Reply:
In Go the "null" value of an integer is `0`. That's what the api is doing.

Reply:
Oh, ok. What happens when you give it a task with a `0` listId or similar things for other items?

Reply:
Not 100% sure but you'll either get an error or some kind of undefined behaviour where the task will be created in the db but you won't be able to access it as it does not belong to a list.

Reply:
Just checked: When you try to set a `list_id` of a task to 0 it will be overridden with the actual id of that task.

Reply:
So should just change the default id to 0 or should I make it nullable and add the null checks?

Reply:
Make it default to 0 and leave it non-nullable.